### PR TITLE
e2e tests for managing vsphere virtual machines

### PR DESCRIPTION
**Pull request by shellymiron**
This PR present the e2e/molecule tests written by QEs (@smiron @epassaro) to manage e2e flows using the validated content roles. in the future those will run in a proper CI.

Molecule is an ansible-testing framework, that allows to write playbooks to validate/assert expected results of a role output.
The structure of the folder will be as follow:
* each role has a testing folder, those testing folders called "Molecule scenario"  
* each testing folder will contain: vars, tasks, post_validations and scenario (if needed) folders. 
* post_validations tests can be written as an ansible playbook or as a pytest file (currently written as a playbook).
* each testing folder/Molecule scenario will run twice (molecule use idempotence mechanism)
* makefile is created to run the molecule test per scenario including all the necessary dependencies

### CODE REVIEW

File: tests/e2e/molecule/manage_vm/scenarios/update_vm.yml [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+# - name: Install VMware Tools on VM with ID {{ vm_moid }}
+#   ansible.builtin.include_tasks:
+#     file: provision_vm/tasks/install_vmware_tools.yml
+
+- name: Test changing state of existing VM {{ vm_update_name }}
+  loop: "{{ vm_states }}"
+  ansible.builtin.include_tasks:
+    file: manage_vm/tasks/update_vm_state.yml
+
+- name: Add Management Network VM Portgroup to all hosts in a cluster
```
Comment (mikemorency) [inline tests/e2e/molecule/manage_vm/scenarios/update_vm.yml:39]: I think this task title should be updated since its not modifying the management network

File: tests/e2e/molecule/manage_vm/scenarios/update_vm.yml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  community.vmware.vmware_portgroup:
+    hostname: "{{ vCenter_hostname }}"
+    username: "{{ vCenter_username }}"
+    password: "{{ vCenter_password }}"
+    validate_certs: "{{ provision_vm_validate_certs }}"
+    cluster_name: "{{ provision_vm_cluster }}"
+    switch: "{{ vswitch_name }}"
+    portgroup: "{{ portgroup_name }}"
+    security:
```
Comment (mikemorency) [inline tests/e2e/molecule/manage_vm/scenarios/update_vm.yml:48]: these two security settings make the network weaker to attacks. I could be wrong, but I dont think they are needed for a regular vm network

File: tests/e2e/molecule/vault_files/eco_vsphere_creds.yml
Comment (mikemorency) [inline tests/e2e/molecule/vault_files/eco_vsphere_creds.yml:1]: does this vault password get supplied via a github actions secret? 
maybe if the file is not too large or doesnt change a lot, we should encrypt the values so people can still see what variables are being supplied?

File: tests/integration/requirements.yml
Comment (mikemorency) [inline tests/integration/requirements.yml:1]: did you mean to change this? seems unrelated to the e2e tests

File: tests/e2e/molecule/README.MD [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+- [Python 3](https://www.python.org/)
+- [Molecule](https://molecule.readthedocs.io/)
+- [ansible-junit](https://github.com/hspaans/ansible-junit) (for JUnit XML output)
+- [pyvmomi](https://github.com/vmware/pyvmomi) (required for vSphere testing)
+
+Additionally, create a password file for the vSphere credentials (vault password file). The password file should contain the vault password for the `vSphere creds` vault. The vault file should include the following variables:
+
+```yaml
+# vault.yml
+vCenter_hostname: "your_vsphere_hostname"
```
Comment (machacekondra) [inline tests/e2e/molecule/README.MD:20]: Is there any reason to use capital letter in the name? Those won't pass ansible-lint by default. I think it's better to not use those, unless there is a really strong reason to do so.

File: tests/e2e/molecule/vault_files/eco_vsphere_creds.yml
Comment (machacekondra) [reply tests/e2e/molecule/vault_files/eco_vsphere_creds.yml:1]: Agree. Either you can use just normal variables, or store it in GH secret.

File: tests/e2e/molecule/README.MD (hunk 3 above)
Comment (elsapassaro) [reply tests/e2e/molecule/README.MD]: thanks, updated to use only lowercase

File: tests/e2e/molecule/manage_vm/scenarios/update_vm.yml (hunk 2 above)
Comment (elsapassaro) [reply tests/e2e/molecule/manage_vm/scenarios/update_vm.yml]: thanks, I checked they are indeed not needed. I removed them

File: tests/e2e/molecule/manage_vm/scenarios/update_vm.yml (hunk 1 above)
Comment (elsapassaro) [reply tests/e2e/molecule/manage_vm/scenarios/update_vm.yml]: Updated to `Add port group with network label {{ portgroup_name }}`

File: tests/integration/requirements.yml
Comment (elsapassaro) [reply tests/integration/requirements.yml:1]: right, I changed it back

File: tests/e2e/molecule/vault_files/eco_vsphere_creds.yml
Comment (elsapassaro) [reply tests/e2e/molecule/vault_files/eco_vsphere_creds.yml:1]: thanks, I removed this file from the repo. It  contained the credentials for the vcenter, as explained in the README [here](https://github.com/redhat-cop/cloud.vmware_ops/pull/23/files#diff-b59d4a9f27f454af561e11816d1c2bad105f8007c4967dec36a22bd58877b4b1R16-R24)

File: tests/e2e/molecule/manage_vm/vars/vars.yml
Comment (mikemorency) [inline tests/e2e/molecule/manage_vm/vars/vars.yml:1]: These variables explicitly set the datastore or host, usually to datastore1. We should be careful not to specify the same host a lot so one doesnt get overloaded. Not a concern right this second I think unless we are going to run a bunch of these tests in parallel. But as we add more test it will definitely be an issue.

If the tests dont need to be on a specific host, we could use `autoselect_datastore` for disk definitions and an NFS share for ISO paths (the iso path needs to be accessible from all possible hosts, so some type of shared storage is required)

File: tests/e2e/molecule/manage_vm/vars/vars.yml [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+provision_vms_template:
+  - provision_vm_name: "qe-provision-vm-template"
+    provision_vm_state: "poweredoff"
+    provision_vm_cdrom:
+      - controller_number: 0
+        unit_number: 0
+        state: present
+        type: iso
+        iso_path: "[datastore1] ISO/rhel-9.3-x86_64-dvd.iso"
+    provision_vm_networks:
```
[the remark this answers is not in the document]
Comment (shellymiron) [reply tests/e2e/molecule/manage_vm/vars/vars.yml:21]: ack I can try to ask that from IT

File: tests/e2e/molecule/manage_vm/vars/vars.yml
Comment (mikemorency) [reply tests/e2e/molecule/manage_vm/vars/vars.yml:1]: Yea i think we should make that a separate ticket/PR. We can always switch to a different host (im sure everyone "defaults" to just using the first datastore/host) if we run into issues before shared storage is set up

**Review by mikemorency — APPROVED** [review]

**Comment by bardielle** [comment]
Please add a changelog file - the name should start with the PR name

**Comment by shellymiron** [comment]
@bardielle can you check if I add it correctly?

File: changelogs/fragments/e2e-manage-vm-tests.yml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+major_changes:
+  - manage_vm - adding manage vm e2e tests using molecule testing framwork. this includes
+                tests for update, delete and adding a vm to the vcenter and validates the outcome of the changes.
+title: provision_vm_test
```
Comment (bardielle) [inline changelogs/fragments/e2e-manage-vm-tests.yml:5]: The file name should start with the PR number 23-e2e-manage-vm-tests.yml

Regarding the title I'm not sure if necessary

Comment (shellymiron) [reply changelogs/fragments/e2e-manage-vm-tests.yml]: tnx! checking

**Review by bardielle — APPROVED** [review]

**Comment by bardielle** [comment]
@shellymiron please rebase on main
